src/test/test-capability.c

```diff
@@ -254,6 +254,13 @@ static void test_capability_get_ambient(void) {
 
         ASSERT_OK(capability_get_ambient(&c));
 
+        r = prctl(PR_CAPBSET_READ, CAP_MKNOD);
```

Comment:
this prctl is not supposed to fail, i'd place `assert_se(r >= 0);` here

Reply:
Maybe, but the current test wasn't supposed to fail either, so I prefer to play it safe rather than having to spend another morning reproducing more weird failures. This is not a test for prctl anyway

Reply:
i disagree. tests are supposed to be tests. if there are unexpected results we want to know them, not tape over them. otherwise tests serve no use.

if you are not interested in reproducing weird failures, please instead just remove the whole test system.

Reply:
and it *definitely* is a test for capabilities, and our expectations around them how they work.

Reply:
> i disagree. tests are supposed to be tests

Unit tests are supposed to be unit tests, ie: test the unit of code associated with it, in this case the helper functions in `src/basic/capability-util`. Everything else is setup and accessory to that, like whether it runs as root or not, etc.

> and it definitely is a test for capabilities, and our expectations around them how they work.

This is a test for `src/basic/capability-util`, not for glibc.
If you want to write tests for `prctl`, please do that on https://sourceware.org/git/?p=glibc.git;a=summary as that's where it's implemented, and where unit tests for its APIs are located

Reply:
i really don't get you. on one hand you want to run the tests on weirdest shittiest most useless test environments somehow. on the other hand you really just want to tape over any unexpected errors as much as you can, and are genuinely disinterested in dealing with cases where our expectations don't hold. i my mind this does not compile. either you are interested in whether our tests work or you aren't, but otherwise this seems like you just want to check some boxes where you can claim the binary starts without it running any useful tests.

i mean, it is our expectation the prctl() works here, and if it doesn't then really really weird stuff is going on and we *want* to know about it. if this fails, then this is no reason to just say "hey, whatever, it's just a test", because then you didn't actually write a useful test as all...

and it's not entirely impossible that our prctl() will fail here, because of seccomp, precisely because containers are doing weird things. but we really should know about those weird things.

Reply:
> i really don't get you. on one hand you want to run the tests on weirdest shittiest most useless test environments somehow.

The report comes from _someone else_ https://github.com/systemd/systemd/issues/35552 I am simply fixing it

> i my mind this does not compile.

It's a very simple concept: unit tests are for testing the unit of code they are attached to, not the random environment they might be executed in. Unit tests do not dictate where they are run, they detect and adapt (with clear outputs/results reflecting that!). The runtime/production/main/whatever_you_want_to_call_them binaries can set expectations on the environment they run in, as can the integration tests, or the upstream CI jobs. The unit tests cannot. The unit tests can check if the preconditions are met - running as root, with permission to create namespaces, with system calls unblocked, and continue if the conditions are met, OR SKIP GRACEFULLY WITH CLEAR RESULTS IF THEY ARE NOT. And yes, system calls that increase the attack surface on the kernel from untrusted foreign build payloads running in containers being blocked is just as valid as a setup condition as not running with root, _FOR A UNIT TEST_.

> either you are interested in whether our tests work or you aren't, but otherwise this seems like you just want to check some boxes where you can claim the binary starts without it running any useful tests.

For the millionth time, no! _IF_ the _conditions_ are met, _then_ the unit of code being tested by the unit tests needs to work, and fail hard if not. If the conditions are not met, THE TEST MUST NOT FAIL, IT MUST SKIP, with a clear skip message, and being marked as skipped, of course, to avoid ambiguity. This is true for root permissions, and for caps, and for a bunch of other stuff, that you already accept without issues.

> and it's not entirely impossible that our prctl() will fail here, because of seccomp, precisely because containers are doing weird things. but we really should know about those weird things.

That's the entire point. The test MUST NOT FAIL if seccomp doesn't let it run, just like it must not fail if it's ran without root, just like it must not fail if it's ran without capabilities, just like it must not fail if it's ran without user namespaces, etc etc etc. _IF_ it has permission to do everything it needs to, and the units of code under test fail, _THEN_ sure, it must fail.

Like it or not, the entire world out there uses docker, that's the reality we live in. Letting tests randomly fail because they run in a container system instead of another, and not because the code being tested has any actual issue, is a bug.
Build environments and runtime environments do not have to match 1:1, and in fact in 99% of the cases (essentially everything apart from your laptop) will differ wildly. Just because a compilation process runs in a certain environment, it does not mean systemd as pid1 will run in the same environment. Build environments are getting more and more locked down, clamping on anything that might allow an escape, as they run a gazillion of external, fundamentally untrusted payloads, so they have to defend themselves, as privilege escalation will give access to _other_ build processes, with the consequences that you might very well imagine.

Reply:
dude, i just want to *know* if weird environments behave in unexpected ways. it's valuable info, and is something we should get to know about, and not systematically ignore.

Reply:
Why would you even want to know such things about weird build environments? Unit tests run at build time, in build environments. They do not run at runtime, in runtime environments. What you want is integration tests to report weird stuff, not unit tests. It's completely useless to know that such and such build system sets seccomp filters. Who cares? It is of no consequence whatsoever.

Reply:
our tests don't run in a vacuum, but on a test environment. we want to know about that and if it doesn't match expectations we make because otherwise our tests are useless and bring no value.

but i really had enough of this. merge this if you want, i really don#t care anymore. i think this discussion is super dumb but it's sooo draining to discuss these things with you.

Reply:
> our tests don't run in a vacuum, but on a test environment. we want to know about that and if it doesn't match expectations we make because otherwise our tests are useless and bring no value.

No, we don't want to know anything about build environments, in fact the less we know, the better. We want to know about runtime environments. Those two things are separate, distinct, disconnected and unrelated. Builds DO NOT HAPPEN in test/runtime environments, this is true in pretty much 99.99% of the cases, the only exception is basically your laptop, I am not aware of a single distribution or production build system where the build environment is the same as the runtime environment. I am pretty sure nobody in their right mind runs a production build system that builds random packages with uid 0 in the root namespace on baremetal without sandboxing.

Tests are not useless just because "make check" runs without root. Maybe some tests are skipped in some environments because they can only run with root, but so what? They still provide great value, for example answering questions such as: is the compiler producing working code?
Full coverage is ensured in our upstream CI, where _we_ control the environment, and don't annoy anybody with these random, unrelated and unsolvable failures.

> but i really had enough of this. merge this if you want, i really don#t care anymore. i think this discussion is super dumb but it's sooo draining to discuss these things with you.

the feeling is very much mutual

Reply:
i vehemently disagree with this BS, but whatever. "the less we know, the better". i mean, come on.

let's maybe remove the test suite altogether if that's the philosophy.

Reply:
Let's maybe make uid 0 in the root namespace on baremetal without sandboxing mandatory for compiling if that's the philosophy. It's very easy to do reductio ad absurdum, but not very useful though